### ISSUE TITLE
Feature Request: Http Client Channel Pool Metrics

### ISSUE THREAD

**Issue by dhoff1985** [micronaut-projects/micronaut-micrometer#26]
It would be nice to have visibility into what is happening in the channel pools of http clients. For example the number of channels checked out, pending acquires, rate/number of new connections created, timings on how long it took to open a new connection, etc.

**Comment by eutkin** [comment]
Yes, this functionality is very much needed. We are seeing performance issues related to connection pooling

**Comment by jjathman-ch** [comment]
@graemerocher I realize this issue is quite old, do you know if an implementation of this would be relatively straightforward for someone to contribute it? I'd be happy to try and take this on but I'm a bit worried the HTTP connection pooling code is too complex for someone outside the Micronaut core team. 

Our main use case is we would like to understand better if we are maxing out our HTTP connection pools or not but as far as I know there's no way to know that currently.

**Comment by graemerocher** [comment]
@yawkat would know more

**Comment by yawkat** [comment]
Depends on the metrics. imo ConnectionManager is fairly clean nowadays, and I think the "state" metrics (e.g. number of open connections) should be fairly easy with access to PoolResizer. Timing metrics may be more difficult because the code is very asynchronous

**Comment by jjathman-ch** [comment]
@yawkat thank you. Do you think updating the state of the metrics within the `PoolResizer#doSomeWork` method would be a reasonable approach? I think just these metrics would be enough for me:

```
pendingRequestCount
pendingConnectionCount 
http1ConnectionCount
http2ConnectionCount
```

**Comment by yawkat** [comment]
sure

**Comment by jjathman-ch** [comment]
Hi, I'm still looking into this feature but wanted to ask for a bit of clarity about how to get started.

From what I can tell, `DefaultHttpClient` has a reference to a `ConnectionManager` instance, which unfortunately is not injected and it just creates on its own. Each `ConnectionManager` will have one or more `Pool` instances which are stored within an instance variable map.

I think ultimately to provide metrics I need to be able to interact with each Pool instance and retrieve some data from it. I think getting the data from each Pool instance will require some changes to its API but that's OK.

The harder problem seems to be how am I able to get a list of all Pool instances to be able to bind metrics to them. I believe the number of pools can change over time so it can't be something that is done at startup of the app. I think either the pool instances need to be registered with the Micronaut application context or some kind of a static field of ConnectionManager which has a list of every pool created by the entire app? @yawkat or @graemerocher any thoughts on if one of these options is better or possible? Or something else I'm missing?
